Source/WebKit/WebProcess/WebPage/WebPage.cpp

```diff
@@ -1092,6 +1094,8 @@ WebPage::~WebPage()
         m_footerBanner->detachFromPage();
 #endif
 
+    WebStorageNamespaceProvider::getOrCreate(*m_pageGroup)->decrementUseCount(sessionStorageNamespaceIdentifier());
```

Comment:
Seems a little weird that we're calling something that might "create" a provider, just to decrement something.

Reply:
I'll respond to the other comments tomorrow, but this is the last improvement that I'm still thinking about. I don't really like this, either, but I'm not yet sure which alternative we should use. Originally, I naively tried `static_cast<WebStorageNamespaceProvider&>(m_page->storageNamespaceProvider()).decrementUseCount(sessionStorageNamespaceIdentifier());`, but that doesn't work because `m_page` is destroyed before we enter the WebPage's destructor. We could let WebPage hold its own reference to the StorageNamespaceProvider, but that's also not ideal. We could also create a static `decrementUseCount` method that operates like the current `getOrCreate()->decrement...()` call, but only if the Namespace currently exists. I would prefer symmetric `increment` and `decrement` functions, but incrementing the useCount seems to be much simpler than decrementing.